include/material/mohr_coulomb.tcc

```diff
@@ -56,311 +55,332 @@ bool mpm::MohrCoulomb<Tdim>::compute_elastic_tensor() {
+  double softening_trial = 0.;
+  //if(yield_type != 0){
+  // Compute plastic multiplier from the current stress state
+  this->compute_df_dp(yield_type, j2,j3,rho,theta,stress,dF_dSigma, dP_dSigma, epds, softening, ptr);
+  // Check the epds
+  if(epds_last < peak_EPDS_ && epds > peak_EPDS_)softening = 0;
+  lambda = dF_dSigma.dot(this->de_ * dstrain) / ((dF_dSigma.dot(this->de_ * dP_dSigma)) + softening);
+  //}
+  // Compute the trial stress
+  //if(yield_type == 0){
+  j2 = j3 = rho = theta = 0.;
+  this->compute_rho_theta(trial_stress, j2, j3, rho, theta);
+  epsilon = (1. / sqrt(3.)) * (trial_stress(0) + trial_stress(1) + trial_stress(2));
+  yield_type_trial = this->check_yield(yield_function_trial, epsilon, rho, theta);
+    //if(yield_type_trial !=0){
+  this->compute_df_dp(yield_type_trial, j2,j3,rho,theta,trial_stress,dF_dSigma_trial, dP_dSigma_trial, epds, softening_trial, ptr);
+    //}
+    // Check the epds
+  if(epds_last < peak_EPDS_ && epds > peak_EPDS_)softening_trial = 0;
+  if(yield_type_trial == 1)
+    lambda_trial = yield_function_trial(0) / ((dF_dSigma_trial.transpose() * de_).dot(dP_dSigma_trial.transpose()) + softening_trial);
   else
-    yield_state = false;
-
-  // compute plastic multiplier from the current stress state
-  this->softening_ = 0;
-  this->compute_df_dp(stress);
-  double lambda = df_dsigma_.dot(this->de_ * dstrain) /
-                  ((df_dsigma_.dot(this->de_ * dp_dsigma_)) + softening_);
-  if (!yield_state) lambda = 0.;
-
-  // compute the trial stress
-  // sigma_trial = sigma + De * dstrain
-  j2_ = j3_ = rho_ = theta_ = 0.;
-  Vector6d trial_stress = stress + (this->de_ * dstrain);
-  this->compute_rho_theta(trial_stress);
-  epsilon =
-      (1. / sqrt(3.)) * (trial_stress(0) + trial_stress(1) + trial_stress(2));
-  double yield_func_trial =
-      sqrt(3. / 2.) * rho_ *
-          ((sin(theta_ + ONETHIRDPI) / (sqrt(3.) * cos(phi_))) +
-           (cos(theta_ + ONETHIRDPI) * tan(phi_) / 3.)) +
-      (epsilon / 3.) * tan(phi_) - c_;
-
-  bool yield_state_trial;
-  if (yield_func_trial > 1.E-22)
-    yield_state_trial = true;
-  else
-    yield_state_trial = false;
-
-  softening_ = 0;
-  this->compute_df_dp(trial_stress);
-  double lambda_trial =
-      yield_func_trial /
-      ((df_dsigma_.transpose() * de_).dot(dp_dsigma_.transpose()) +
-       softening_);
-
-  double p_multiplier;
-  if (yield_state)
-    p_multiplier = lambda;
-  else if (!yield_state) {
-    if (yield_state_trial)
+    lambda_trial = yield_function_trial(1) / ((dF_dSigma_trial.transpose() * de_).dot(dP_dSigma_trial.transpose()) + softening_trial);
+  // Compute the correction stress
+  double p_multiplier = 0.;
+  Vector6d dP_dSigma_Final = Vector6d::Zero();
+  bool PDS_update = false;
+  // check the load and unload
+  double dyfun_t = yield_function_trial(0) - yield_function(0);
+  double dyfun_s = yield_function_trial(1) - yield_function(1);
+  if(yield_type !=0){
+    if(yield_type == 1 && dyfun_t > 0){
```

Comment:
It looks both `if` statements do the same thing? Can we instead have it inside one `if` statement?